**Introduction to NumPy**

NumPy is the fundamental package for scientific computing in Python. 

NumPy was created in 2005 by Travis Oliphant. It is an open source project and you can use it freely.

NumPy stands for Numerical Python.

It is a Python library that provides a multidimensional array object, various derived objects (such as masked arrays and matrices), and an assortment of routines for fast operations on arrays, including mathematical, logical, shape manipulation, sorting, selecting, I/O, discrete Fourier transforms, basic linear algebra, basic statistical operations, random simulation and much more.

At the core of the NumPy package, is the ndarray object. This encapsulates n-dimensional arrays of homogeneous data types, with many operations being performed in compiled code for performance. There are several important differences between NumPy arrays and the standard Python sequences:

NumPy arrays have a fixed size at creation, unlike Python lists (which can grow dynamically). Changing the size of an ndarray will create a new array and delete the original.

The elements in a NumPy array are all required to be of the same data type, and thus will be the same size in memory. The exception: one can have arrays of (Python, including NumPy) objects, thereby allowing for arrays of different sized elements.

NumPy arrays facilitate advanced mathematical and other types of operations on large numbers of data. Typically, such operations are executed more efficiently and with less code than is possible using Python’s built-in sequences.

A growing plethora of scientific and mathematical Python-based packages are using NumPy arrays; though these typically support Python-sequence input, they convert such input to NumPy arrays prior to processing, and they often output NumPy arrays. In other words, in order to efficiently use much (perhaps even most) of today’s scientific/mathematical Python-based software, just knowing how to use Python’s built-in sequence types is insufficient - one also needs to know how to use NumPy arrays.

**Why Use NumPy?**

- In Python we have lists that serve the purpose of arrays, but they are slow to process.

- NumPy aims to provide an array object that is up to 50x faster than traditional Python lists.

- The array object in NumPy is called ndarray, it provides a lot of supporting functions that make working with ndarray very easy.

- Arrays are very frequently used in data science, where speed and resources are very important.

**Python Objects**

1. high-level number objects: integers,floating point
2. containers: lists(costless insertion and append), dictionaries(fast lookup)

**Why is NumPy Faster Than Lists?**

NumPy arrays are stored at one continuous place in memory unlike lists, so processes can access and manipulate them very efficiently.

This behavior is called locality of reference in computer science.

This is the main reason why NumPy is faster than lists. Also it is optimized to work with latest CPU architectures.

**NumPy is useful:**

- useful for array oriented computing
- efficiency as good as hardware (closer to hardware efficiency)
- extension package to Python for multi-dimensional arrays
- developed for faster scientific computations 


**Which Language is NumPy written in?**

NumPy is a Python library and is written partially in Python, but most of the parts that require fast computation are written in C or C++.

**Getting started with NumPy**

Importing NumPy



In [ ]:
import numpy

**Creating an array using NumPy**

In [4]:
import numpy

ar = numpy.array([21, 31, 41, 51,61])

print(ar)

[21 31 41 51 61]


**NumPy as np**

**NumPy** is usually imported under the **np** alias.

```
alias: In Python alias are an alternate name for referring to the same thing.
```





Create an alias with the as keyword while importing:

In [5]:
import numpy as np

ar = np.array([21, 31, 41, 51,61])

print(ar)

[21 31 41 51 61]


In [7]:
# Checking NumPy Version

print(np.__version__)

1.21.6


**Creating numPy array of all the numbers between 0 to 20**

In [9]:
import numpy as np

a = np.arange(20)
print(a)


[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19]


**NumPy is FASTER: let's prove it**

**Using Python List**

In [10]:
s = range(200)

# computing the time consumed for the computation

%timeit [i**2 for i in s]

64.3 µs ± 18.5 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


**Using NumPy**

In [13]:
p = np.arange(200)

# computing the time consumed for the computation

%timeit p**2

1.18 µs ± 382 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


We can clearly see that NumPy has taken only 1.18 µs while python has taken 64.3 µs which is nearly 54 times more than what NumPy has taken to compute.

**Creating 1-dimensional arrays**

In [2]:
import numpy as np

ar = np.array([10, 20, 30, 40, 50])

print(ar)

[10 20 30 40 50]


**Finding dimension of an array**

In [3]:
ar.ndim

1

**Finding shape of an array**

In [4]:
ar.shape

(5,)

**Finding length of an array**

In [6]:
len(ar)

5

**2-D Arrays**


---


An array that has 1-D arrays as its elements is called a 2-D array.

These are often used to represent matrix or 2nd order tensors.

**Creating 2-D array**

Create a 2-D array containing two arrays with the values 17,27,37,47 and 12,15,16,18:

In [7]:
import numpy as np

ar = np.array([[17,27,37,47], [12,15,16,18]])

print(ar)

[[17 27 37 47]
 [12 15 16 18]]


In [8]:
# dimension of an array

ar.ndim

2

In [9]:
# shape of an array

ar.shape

(2, 4)

In [10]:
# length of an array 
# returns the length of the first dimension

len(ar)

2

#3-D arrays

An array that has 2-D arrays (matrices) as its elements is called 3-D array.

These are often used to represent a 3rd order tensor.

# Creating 3-D array



In [11]:
import numpy as np

ar = np.array([[[10, 20], [30, 40]], [[10, 20], [30, 40]]])

print(ar)

[[[10 20]
  [30 40]]

 [[10 20]
  [30 40]]]


In [12]:
ar.ndim

3

In [13]:
ar.shape

(2, 2, 2)

In [14]:
# another example
import numpy as np

ar = np.array([[[11, 13, 15], [14, 25, 46]], [[31, 22, 65], [34, 77, 88]]])

print(ar)

[[[11 13 15]
  [14 25 46]]

 [[31 22 65]
  [34 77 88]]]


In [15]:
ar.ndim

3

In [16]:
ar.shape

(2, 2, 3)

In [17]:
len(ar)

2

In [18]:
# Another example
import numpy as np

a = np.array(101)
b = np.array([5, 7, 17, 29, 23])
c = np.array([[21, 42, 53], [34, 55, 90]])
d = np.array([[[111, 312, 234], [89, 7, 4]], [[3, 1, 7], [8, 7, 1]]])

print(a.ndim)
print(b.ndim)
print(c.ndim)
print(d.ndim)

0
1
2
3


# Things to know 

- **1D Array -** Vector

- **2D Array -** Matrix

- **nD Array -** Tensor

where n can be 3,4,5,...

# Few other ways to creating array
 

# using arange function

In [22]:
import numpy as np

ar = np.arange(5)
print(ar)

[0 1 2 3 4]


In [24]:
# using start, end, step
ar = np.arange(2,10,3)
print(ar)

[2 5 8]


In [26]:
# Using linspace - linear space
# start, end, number of points in between

ar = np.linspace(2,3,5)
print(ar)


[2.   2.25 2.5  2.75 3.  ]


# Creating Common Arrays

In [34]:
ar = np.ones((2,2))
print(ar,"\n")

ar1 = np.ones((4,4))
print("ar1 is:\n")
print(ar1)

[[1. 1.]
 [1. 1.]] 

ar1 is:

[[1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]]


In [36]:
ar = np.zeros((2,2))
print(ar,"\n")

ar1 = np.zeros((4,4))
print("ar1 is:\n")
print(ar1)

[[0. 0.]
 [0. 0.]] 

ar1 is:

[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]


# Assignment on NumPy Array

***Create a 2D array with ones on the diagonal and zeores elsewhere.***

# here is the Solution

In [37]:
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

# Creating array using diag function

In [40]:
ar = np.diag([9,8,7,6])
print(ar)

[[9 0 0 0]
 [0 8 0 0]
 [0 0 7 0]
 [0 0 0 6]]


#Extracting the diaggonal of matrix

In [41]:
np.diag(ar)

array([9, 8, 7, 6])

# Creating array using random

In [43]:
ar = np.random.rand(4)  # returns the array of size 4 (4 elements)
print(ar)

[0.65209875 0.83113221 0.59742128 0.38829385]
